ce/api/streamflow/watershed.py

```diff
@@ -0,0 +1,421 @@
+
+from ce.api.geospatial import \
+    geojson_feature, outline_cell_rect, outline_point_buff, WKT_point_to_lonlat
+from ce.api.util import is_valid_index, vec_add, neighbours
+from modelmeta import \
+    DataFile, DataFileVariable, Ensemble, EnsembleDataFileVariables
+
+
+# TODO: Move DataGrid and its exceptions to separate module
+
+class DataGridError(Exception):
+    """Base class for exceptions in this module."""
+    pass
+
+
+class DataGridIndexError(DataGridError):
+    """Exception for attempt to incorrectly index a data grid. This is more
+    stringent than a simple IndexError.
+
+    Attributes:
+        index: offending index
+        shape: offended grid shape
+        message: explanation of the error, generated from params
+    """
+
+    def __init__(self, index, shape):
+        self.index = index
+        self.shape = shape
+        self.message = 'Index {} is not valid for a grid of shape {}'\
+            .format(index, shape)
+
+
+class DataGrid:
+    """Represents the contents of a gridded dataset that are relevant
+    to the watershed endpoint. The existence of this class has two motivations:
+
+    - Factor out common operations on datasets (e.g., convert between lonlat
+        and xy coordinates).
+    - Make it much simpler to construct test data for the `worker` function.
+    """
+
+    def __init__(self, longitudes, latitudes, values, units=None):
+        self.longitudes = longitudes
+        self.latitudes = latitudes
+        self.values = values
+        self.units = units
+        self.lon_step = longitudes[1] - longitudes[0]
+        self.lat_step = latitudes[1] - latitudes[0]
+
+    @staticmethod
+    def from_nc_dataset(dataset, variable_name):
+        """Factory method. Extracts relevant data from a netcdf file (`Dataset`)
+        with standard contents and returns it as a `DataGrid`."""
+        return DataGrid(
+            dataset.variables['lon'],
+            dataset.variables['lat'],
+            dataset.variables[variable_name],
+            dataset.variables[variable_name].units,
+        )
+
+    def is_compatible(self, other):
+        """Return a boolean indicating whether this `DataGrid` and another are
+        compatible. Compatible means that their lon and lat grids have the same
+        step size."""
+        return math.isclose(self.lon_step, other.lon_step) and \
+               math.isclose(self.lat_step, other.lat_step)
+
+    def check_valid_index(self, index):
+        if not is_valid_index(index, self.values.shape):
+            raise DataGridIndexError(index, self.values.shape)
+
+    def lonlat_to_xy(self, lonlat):
+        """Returns the (x, y) data index for a given lon-lat coordinate,
+        switching the order of the coordinates. Checks that the index is
+        valid for the grid; we must at minimum exclude negative index values,
+        which are valid but wrong in this application."""
+        x = int(round((lonlat[1] - self.latitudes[0]) / self.lat_step))
```

Comment:
Maybe that datasets that will be used for the watershed API *are* consistent. But I would recommend at least an `assert` statement ensuring that the deltas are all equal, if the rest of the code depends on that.

Reply:
Hmm, this is a trickier question than it at first seems. It's worth a little analysis:

- The watershed (upstream cells from mouth) is built based in part on the sign of the lat an lon steps (this determines the "direction matrix"). But this is not a compatibility issue.
- Upstream cells are returned as cell indices. They are converted to lonlat pairs for the next steps.
- The elevation and area lists are built by querying on lonlats (converted from upstream cell indices). I don't *believe* this code requires even that elevation and area datasets are on the same grid: For each variable (elevation, area), a lonlat is converted to the nearest cell index and that is used. 
   - Now the question is raised of whether this approximation is acceptable, or whether we must insist that the grids are compatible in the sense that each lonlat specified corresponds exactly to a lonlat in the grid.
- Therefore the comparison of whether lonlat grids are compatible because they have the same step is either
    - irrrelevant, or
    - insufficient; we need to check whether the lonlats are the same points, not just the same distance apart.

@corviday , do you have any thoughts on this?

Reply:
Some further thoughts:

- I imagined the purpose of `DataGrid`, despite its general sounding name, to be to handle only the flow direction, elevation, and area datasets needed in the watershed API. So at present it builds in some assumptions about those datasets inferred from the examples we have. Those assumptions are:
    - Uniformly spaced latitude and longitude grids
    - A loose notion of compatibility between grids, which looks like it probably was too loose (see above)

- If we want `DataGrid` to represent more general datasets, then it will have to drop these assumptions. Furthermore, it probably wouldn't be appropriate to build in an `is_compatible` method, since the compatibility criterion could vary greatly from application to application. Little as I like object oriented code in general, this would be a case for a base class with usage specific subclasses.

- So maybe it needs (a) a different name that implies its specific usage, and (b) subclassing from a more generic class.

Reply:
I refactored DataGrid into a base and VIC/watershed-specific subclass, and added a more stringent compatibility condition, which is that the lat and lon coordinates must be the same values where they overlap spatially.